src/mono/mono/metadata/loader.c

```diff
@@ -81,30 +81,32 @@ static gint32 signatures_size;
 void
 mono_loader_init ()
 {
-	static gboolean inited;
+	static volatile gboolean inited = FALSE; // We make this volatile to avoid race conditions
 
-	// FIXME: potential race
-	if (!inited) {
-		mono_coop_mutex_init_recursive (&loader_mutex);
-		mono_os_mutex_init_recursive (&global_loader_data_mutex);
-		loader_lock_inited = TRUE;
+	if (inited) {
+		/* Do nothing if this function has already been called */
+		return;
+	}
+	
+	inited = TRUE;
 
-		mono_global_loader_cache_init ();
+	mono_coop_mutex_init_recursive (&loader_mutex);
+	mono_os_mutex_init_recursive (&global_loader_data_mutex);
+	loader_lock_inited = TRUE;
```

Comment:
It'd be great if these globals could be statically initialized!
I agree the fix isn't right, because it can call the initializers multiple times concurrently, leaving really totally corrupt data, depending on underlying details..

The mono_once should probably be bought to bear, generally, seriously.